openide.awt/src/org/openide/awt/ContextAction.java

```diff
@@ -181,36 +317,90 @@ public Performer(
             ContextActionEnabler<Data> e
         ) {
             Map<Object, Object> map = new HashMap<Object, Object>();
-            map.put("delegate", p);
-            map.put("enabler", e);
+            map.put("delegate", p); // NOI18N
+            map.put("enabler", e);  // NOI18N
             this.delegate = map;
         }
-
-        @SuppressWarnings("unchecked")
-        public void actionPerformed(
-            ActionEvent ev, List<? extends Data> data, Lookup.Provider everything
-        ) {
-            Object obj = delegate.get("delegate"); // NOI18N
-            if (obj instanceof ContextActionPerformer) {
-                ContextActionPerformer<Data> perf = (ContextActionPerformer<Data>)obj;
-                perf.actionPerformed(ev, data);
-                return;
+        
+        void clear() {
+            stopListeners();
+            Reference r = instDelegate;
+            instDelegate = null;
+            if (r != null) {
+                Object o = r.get();
+                if (o instanceof Performer) {
+                    ((Performer)o).clear();
+                }
             }
-            if (obj instanceof Performer) {
-                Performer<Data> perf = (Performer<Data>)obj;
-                perf.actionPerformed(ev, data, everything);
-                return;
+        }
+        
+        void attach(ContextAction a) {
+            this.owner = new WeakReference<>(a);
+        }
+        
+        /**
+         * Creates a delegate. 
+         * @param everything
+         * @param data
+         * @return 
+         */
+        Object delegate(Lookup.Provider everything, List<?> data) {
+            return delegate0(everything, data, true);
+        }
+        
+        private Object delegate0(Lookup.Provider everything, List<?> data, boolean getAction) {
+            Object d = instDelegate != null ? instDelegate.get() : null;
+            if (d != null) {
+                if (getAction && (d instanceof Performer)) {
+                    return ((Performer)d).delegate0(everything, data, getAction);
+                }
+                return d;
             }
-            if (obj instanceof ContextAwareAction) {
-                Action a = ((ContextAwareAction)obj).createContextAwareInstance(everything.getLookup());
-                a.actionPerformed(ev);
-                return;
+            d = createDelegate(everything, data);
+            if (d != null) {
+                if (getAction && (d instanceof Performer)) {
+                    final Object fd = d;
+                    instDelegate = new WeakReference<Object>(d) { private Object hardRef = fd; };
```

Comment:
What's going on here? Trying to wrap my head around this while debugging NETBEANS-4754...